redisinsight/api/src/modules/ssh/ssh-tunnel.provider.ts

```diff
@@ -1,38 +1,176 @@
 import { HttpException, Injectable } from '@nestjs/common';
+import { Client } from 'ssh2';
+import { createServer, Server, AddressInfo } from 'net';
 import { SshTunnel } from 'src/modules/ssh/models/ssh-tunnel';
-import {
-  UnableToCreateTunnelException,
-} from 'src/modules/ssh/exceptions';
+import { UnableToCreateTunnelException } from 'src/modules/ssh/exceptions';
 import { Endpoint } from 'src/common/models';
 import { SshOptions } from 'src/modules/ssh/models/ssh-options';
-import { createTunnel } from 'tunnel-ssh';
 
 @Injectable()
 export class SshTunnelProvider {
-  public async createTunnel(target: Endpoint, sshOptions: SshOptions) {
-    try {
-      const [server, client] = await createTunnel({
-        autoClose: true,
-      }, {
-        host: '127.0.0.1',
-      },
-      {
-        ...sshOptions,
-      }, {
-        dstAddr: target.host,
-        dstPort: target.port,
+  private activeConnections: Map<string, { client: Client; server: Server | null }> = new Map();
+
+  private getConnectionKey(target: Endpoint, sshOptions: SshOptions): string {
+    return `${sshOptions.host}:${sshOptions.port}-${target.host}:${target.port}`;
+  }
+
+  private cleanupConnection(connectionKey: string): void {
+    const connection = this.activeConnections.get(connectionKey);
+    if (connection) {
+      console.log('Cleaning up connection:', connectionKey);
+    
+      // Clean up server
+      if (connection.server) {
+        connection.server.close();
+        connection.server.unref();
+      }
+    
+      // Clean up client
+      if (connection.client) {
+        connection.client.removeAllListeners(); // Remove all event listeners
+        connection.client.end();
+        connection.client.destroy();
+      }
+    
+      this.activeConnections.delete(connectionKey);
+    }
+  }
+
+  public async createTunnel(target: Endpoint, sshOptions: SshOptions): Promise<SshTunnel> {
+    return new Promise((resolve, reject) => {
+      const connectionKey = this.getConnectionKey(target, sshOptions);
+      
+      // Cleanup any existing connection
+      this.cleanupConnection(connectionKey);
+
+      const client = new Client();
+      let server: Server | null = null;
+
+      client.on('keyboard-interactive', (name, instructions, instructionsLang, prompts, finish) => {
+        finish(['032100']);
```

Comment:
@GnaneshKunal What is this string representing? Short google search shows it can be used as a sort of a 'hack' for dealing with some authenticators.

Reply:
@KIvanow, this PR is more of a POC and approach. 
Ideally, we shouldy implement a UI form elementthat takese in the TOTP andpassess it here in the `finish` function.

Reply:
got it, thanks!